### PR TITLE
Noam/dee 381 create token classification use case

### PR DESCRIPTION
**Pull request by noamzbr**
New token classification dataset & quickstart for token classification.

### CODE REVIEW

File: deepchecks/nlp/datasets/token_classification/scierc_ner.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+# This file is part of Deepchecks.
+# Deepchecks is distributed under the terms of the GNU Affero General
+# Public License (version 3 or later).
+# You should have received a copy of the GNU Affero General Public License
+# along with Deepchecks.  If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+#
+"""Dataset containing abstracts from scientific journals and named entity annotations for relevant scientific words.
+
+The data contains 350 samples for the train set and 100 samples for the test set. Each dataset sample is a tokenized
```
Comment (nirhutnik) [inline deepchecks/nlp/datasets/token_classification/scierc_ner.py:13]: That's the original size? That's really small for us to be able to demonstrate some thing... no? E.g. embeddings drift

File: deepchecks/nlp/datasets/token_classification/scierc_ner.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            include_properties, include_embeddings = False, False
+
+    data = load_all_data()
+    train, test = data['train'], data['test']
+    # Delete the pred field of the dictionary
+    del train['pred']
+    del test['pred']
+
+    if data_format.lower() != 'textdata':
+        return train, test
```
Comment (nirhutnik) [inline deepchecks/nlp/datasets/token_classification/scierc_ner.py:138]: What about data_format == 'dataframe'?

File: deepchecks/utils/builtin_datasets_utils.py [hunk 3]
```diff
@@ -34,10 +38,14 @@ def read_and_save_data(assets_dir, file_name, url_to_file, file_type='csv', to_n
...
+            with open(assets_dir / file_name, 'w', encoding='utf-8') as f:
+                json.dump(data, f)
         else:
             raise ValueError('file_type must be either "csv" or "npy"')
 
-    if to_numpy:
+    if to_numpy and (file_type in {'csv', 'npy'}):
         if isinstance(data, pd.DataFrame):
             data = data.to_numpy()
         elif not isinstance(data, np.ndarray):
```
Comment (nirhutnik) [inline deepchecks/utils/builtin_datasets_utils.py:51]: what if to_numpy and json? Nothing happens? A bit weird. Either should be documented or resolved

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+************************************************
+
+In this quickstart guide, we will give a brief example of using the deepchecks NLP package to analyze and evaluate token
+classification tasks. For a more complete example showcasing the range of checks and capabilities of the NLP package,
+refer to our :ref:`Multiclass Quickstart <nlp__multiclass_quickstart>`. We will cover the following steps:
+
+1. `Creating a TextData object and auto calculating properties <#setting-up>`__
+2. `Running checks <#running-checks>`__
+
+To run deepchecks for NLP, you need the following for both your train and test data:
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:15]: not for NLP, but for token classification

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+1. Your :ref:`text data <nlp__textdata_object>` - a list of strings, each string is a single sample
+   (can be a sentence, paragraph, document, etc.).
+2. Your labels - either a :ref:`Text Classification <nlp_supported_text_classification>` label or a
+   :ref:`Token Classification <nlp_supported_token_classification>` label. These are not needed for checks that
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:20]: Weird to give text classification as an option here

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+   don't require labels (such as the Embeddings Drift check or most data integrity checks), but are needed for
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:21]: ref to check?

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+from deepchecks.nlp import TextData
+from deepchecks.nlp.datasets.token_classification import scierc_ner
+
+train, test = scierc_ner.load_data(data_format='Dict')
+pprint(train['text'][0][:10])
+pprint(train['label'][0][:10])
+
+# %%
+#
+# The SCIERC dataset is a dataset of scientific articles with annotations for named entities, relations and coreference.
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:60]:
```suggestion
# The SCIERC dataset is a dataset of scientific articles with annotations for named entities, relations and coreferences.
```

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+# In this example we'll only use the named entities annotations, which are the labels we'll use for our token
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:61]:
```suggestion
# In this example we'll only use the named entity annotations, which are the labels we'll use for our token
```

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+# classification task.
+# We can see that we have the article text itself, and the labels for each token in the text in the
+# :ref:`IOB format <nlp__supported_token_classification>`.
+#
+# Create a TextData Objects
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:66]:
```suggestion
# Create a TextData Object
```

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+# -------------------------
+#
+# We can now create a :ref:`TextData <nlp__textdata_object>` object for the train and test dataframes.
+# This object is used to pass your data to the deepchecks checks.
+#
+# To create a TextData object, the only required argument is the tokenized text itself, but passing only the text
+# will prevent multiple checks from running. In this example we'll pass the label and define the task type.
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:73]: I wrote that part of the code and I'm not sure I get what you're saying here.
I think this is the place to explain the advantages and disadvantages of raw_text vs tokenized_text, the automatic inference, and maybe a word on the format?

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+# Some of deepchecks' checks use properties of the text samples for various calculations. Deepcheck has a wide
+# variety of such properties, some simple and some that rely on external models and are more heavy to run. In order
+# for deepchecks' checks to be able to access the properties, they must be stored within the
+# :ref:`TextData <nlp__textdata_object>` object. You can read more about properties in the
+# :ref:`Property Guide <nlp__properties_guide>`.
+
+# properties can be either calculated directly by Deepchecks
+# or imported from other sources in appropriate format
+
+# device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:92]: Add a sentence saying that this is how we calculate

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+# train.calculate_builtin_properties(
+#   include_long_calculation_properties=True, device=device
+# )
+# test.calculate_builtin_properties(
+#   include_long_calculation_properties=True,  device=device
+# )
+
+train_properties, test_properties = scierc_ner.load_properties()
```
Comment (nirhutnik) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:100]: Add a sentence saying we already precalculated and now we'll load them

**Review by nirhutnik — APPROVED** [review]

File: tests/nlp/test_datasets.py [hunk 13]
```diff
@@ -82,3 +83,22 @@ def test_just_dance_comment_analysis():
...
+    assert_that(len(train_dict['text']), equal_to(len(train_dict['label'])))
+    assert_that(len(test_dict['text']), equal_to(len(test_dict['label'])))
+
+    assert_that(len(train_ds.text), equal_to(len(train_preds)))
+    assert_that(len(test_ds.text), equal_to(len(test_preds)))
+
+    assert_that(train_props.columns, contains_exactly(*test_props.columns))
+
+    assert_that(train_embeddings.shape, contains_exactly(350, 1536))
+    assert_that(test_embeddings.shape, contains_exactly(100, 1536))
```
Comment (nirhutnik) [inline tests/nlp/test_datasets.py:104]: No test for loading DataFrame :)

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py (hunk 6 above)
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:21]: included later when check is actually shown

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py (hunk 10 above)
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:73]: What I'm actually doing here is telling them that label is not required.
Also for token classification tokenized_text is mandatory, but will try to elaborate.

File: tests/nlp/test_datasets.py (hunk 13 above)
Comment (noamzbr) [reply tests/nlp/test_datasets.py:104]: Can't load as dataframe

File: deepchecks/nlp/datasets/token_classification/scierc_ner.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    include_embeddings : bool, default: True
+        If True, the returned data will include embeddings of the comments. Incompatible with data_format='DataFrame'
+
+    Returns
+    -------
+    train, test : Tuple[Union[TextData, Dict]
+        Tuple of two objects represents the dataset split to train and test sets.
+    """
+    if data_format.lower() not in ['textdata', 'dict']:
+        raise ValueError('data_format must be either "Dataset" or "Dict"')
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/datasets/token_classification/scierc_ner.py:124]: 1. Right, fixed in all 3
2. Don't think so, it's not as part of a check
3. No dataframe for this task (token classification doesn't work with dataframe so well)

File: deepchecks/nlp/datasets/token_classification/scierc_ner.py (hunk 2 above)
Comment (noamzbr) [reply deepchecks/nlp/datasets/token_classification/scierc_ner.py:138]: not an option

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py (hunk 11 above)
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:92]: Don't think it's needed, should be pretty clear for most nlp users and it's long enough already. Added a line for that in the property guide as it was missing.

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+# %%
+
+from deepchecks.nlp.checks import TextEmbeddingsDrift
+
+check = TextEmbeddingsDrift()
+res = check.run(train, test)
+res.show()
+
+# %%
+# The check shows the samples from the train and test datasets as points in the 2-dimensional reduced embedding
```
Comment (shir22) [inline docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:170]: do we want to add a link to it (the check)?
(cause as a reader I wonder "how is the dimension reduced" etc, and don't think it's relevant here but would want to easily read further

**Review by shir22 — APPROVED** [review]

File: deepchecks/nlp/metric_utils/token_classification.py [hunk 16]
```diff
@@ -32,15 +32,15 @@ def get_scorer_dict(
     """Return a dict of scorers for token classification.
 
     Parameters
-    -----------
+    ----------
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/metric_utils/token_classification.py:35]: it was a doc build warning so did it

File: tests/nlp/test_datasets.py (hunk 13 above)
Comment (noamzbr) [reply tests/nlp/test_datasets.py:104]: doesn't, fixed it a while ago

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    import sys
+    !{sys.executable} -m pip install deepchecks[nlp] -U --quiet #--user
+
+Some properties calculated by ``deepchecks.nlp`` require additional packages to be installed. You can
+install them by running:
+
+.. code:: python
+
+    import sys
+    !{sys.executable} -m pip install [nlp-properties] -U --quiet #--user
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:40]: fixed the initial issue, but yes need to install both for all properties ot run

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+train = TextData(tokenized_text=train['text'], label=train['label'], task_type='token_classification')
+test = TextData(tokenized_text=test['text'], label=test['label'], task_type='token_classification')
+
+# %%
+# Calculating Properties
+# ----------------------
+#
+# Some of deepchecks' checks use properties of the text samples for various calculations. Deepcheck has a wide
+# variety of such properties, some simple and some that rely on external models and are more heavy to run. In order
+# for deepchecks' checks to be able to access the properties, they must be stored within the
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:87]: Tried a better phrasing

File: docs/source/nlp/tutorials/quickstarts/plot_token_classification.py (hunk 15 above)
Comment (noamzbr) [reply docs/source/nlp/tutorials/quickstarts/plot_token_classification.py:170]: There's one above

File: deepchecks/nlp/datasets/token_classification/scierc_ner.py [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+    train_properties = read_and_save_data(ASSETS_DIR, 'train_properties.csv', _TRAIN_PROP, to_numpy=False,
+                                          include_index=False)
+    test_properties = read_and_save_data(ASSETS_DIR, 'test_properties.csv', _TEST_PROP, to_numpy=False,
+                                         include_index=False)
+    return train_properties, test_properties
+
+
+def load_data(data_format: str = 'TextData', include_properties: bool = True, include_embeddings: bool = False) -> \
+        t.Tuple[t.Union[TextData, pd.DataFrame], t.Union[TextData, pd.DataFrame]]:
+    """Load and returns the SCIERC Abstract NER dataset (token classification).
```
[the remark this answers is not in the document]
Comment (noamzbr) [reply deepchecks/nlp/datasets/token_classification/scierc_ner.py:105]: will consider in seperate pr

**Pull request merged**